pkg/export/setup/setup.go

```diff
@@ -82,138 +83,147 @@ func SetGlobal(exporter *export.Exporter) (err error) {
 // Global returns the global instance of the GCM exporter.
 func Global() *export.Exporter {
 	if globalExporter == nil {
-		// This should usually be a panic but we set an inactive default exporter in this case
-		// to not break existing tests in Prometheus.
+		if !testing.Testing() {
+			panic("must set a global exporter")
+		}
+
 		fmt.Fprintln(os.Stderr, "No global GCM exporter was set, setting default inactive exporter.")
+
+		// We don't want to change all upstream Prometheus unit tests, so let's just create a
+		// disabled exporter. These are created on-demand to prevent race conditions between tests.
 		return export.NopExporter()
 	}
 	return globalExporter
 }
 
-// FromFlags returns a constructor for a new exporter that is configured through flags that are
-// registered with the given application. The constructor must be called after the flags
-// have been parsed.
-//
-// NOTE(bwplotka): This method should only setup flags, no extra logic should be done here
-// as we don't have a logger ready, and nothing was logged for the binary yet.
-// Potential risky logic can be moved to the returned function we return here.
-// See b/344740239 on how hard is to debug regressions here.
-func FromFlags(a *kingpin.Application, userAgentProduct string) func(context.Context, log.Logger, prometheus.Registerer) (*export.Exporter, error) {
-	var (
-		metadataFetchTimeout time.Duration
-		opts                 export.ExporterOpts
-	)
-	opts.UserAgentProduct = userAgentProduct
-
+func ExporterOptsFlags(a *kingpin.Application, opts *export.ExporterOpts) {
```

Comment:
suggestion: It feels brittle to rely on `opts` being `.Default()`ed before this is called every time.

What if we did the defaulting of `opts` in this routine, like before? Tying the flag/config definitions and default behavior centrally makes a lot of sense to me.

I think this also comes back to my previous comment of around how we envision clients handling the `ExporterOpts` struct.

Reply:
Addressing in https://github.com/GoogleCloudPlatform/prometheus-engine/pull/1001